### DISCUSSION TITLE
Enhance AppProps typing

### DISCUSSION THREAD

**Discussion by MaximeBernard** [vercel/next.js#18085] [Ideas]
# Feature request

When overriding `_app.tsx`, I'd like to have cleaner typings for my `AppProps`.

Given the following sample 
```typescript
export default function MyApp({ Component, pageProps }: AppProps): ReactElement {
  // pageProps: any
  return <Component {...pageProps} />
}
```

## Concern 1
My first concern is `any` should never be used - `unknown` should be preferred.

(TypeScript documentation)
> After all, remember that all the convenience of any comes at the cost of losing type safety. Type safety is one of the main motivations for using TypeScript and you should try to avoid using any when not necessary.

### Solution

```typescript
export declare type AppProps = Pick<PrivateRouteInfo, 'Component' | 'err'> & {
    router: Router;
} & Record<string, unknown>;
```

## Concern 2
In some cases, we actually know what will be in these props and we should be able to type it.

Use case
```typescript
export default function MyApp({ Component, pageProps }: AppProps): ReactElement {
  return <Component {...pageProps} />
}
type AppInitialProps = {
  pageProps: { lang:'en' | 'pt' | 'fr' };
};
MyApp.getInitialProps = (appContext: AppContextType): AppInitialProps => {
  const lang = 'en';
  return { pageProps: { lang } };
};
```

### Solution 
```typescript
export declare type AppProps<T = Record<string, unknown>> = Pick<PrivateRouteInfo, 'Component' | 'err'> & {
    router: Router;
} & T;
```

### Use-case with solution
```typescript
export default function MyApp({ Component, pageProps }: AppProps<AppInitialProps>): ReactElement {
  // pageProps: { lang:'en' | 'pt' | 'fr' }
  return <Component {...pageProps} />
}
type AppInitialProps = {
  pageProps: { lang:'en' | 'pt' | 'fr' };
};
MyApp.getInitialProps = (appContext: AppContextType): AppInitialProps => {
  const lang = 'en';
  return { pageProps: { lang } };
};
```

## Concern 3
This is some "could be even better" thing. I found `AppInitialProps` in `utils.d.ts` that looks like this:
```typescript
export declare type AppInitialProps = {
    pageProps: any;
};
```
Applying my previous concerns, I think it would be better if it looked like 
```typescript
export declare type AppInitialProps<T = unknown> = {
    pageProps: T;
};
```
I wouldn't even have to declare my own `AppInitialProps` and just use `next` typing as
```typescript
MyApp.getInitialProps = (appContext: AppContextType): AppInitialProps<{ lang:'en' | 'pt' | 'fr' }> => {
  const lang = 'en';
  return { pageProps: { lang } };
};
```

**Comment by usagi-f** [comment]
I agree with. I think the next dev team is worth considering this.
